### DISCUSSION TITLE
mps-platform-docs/platform_essentials/accessing_the_project/

### DISCUSSION THREAD

**Discussion by giscus** [mbeddr/mps-platform-docs#15] [General]
# mps-platform-docs/platform_essentials/accessing_the_project/

This website collects practical knowledge about MPS and popular MPS platforms such as MPS-extensions and mbeddr. It is written mainly in the form of a Q&A.

http://mbeddr.com/mps-platform-docs/platform_essentials/accessing_the_project/

**Comment by BWeng20** [comment]
A word about "Typically, there also should not be a need to access the project in the generator."

Yes, we need. And I guess the use-case is not so uncommon. 

The output of our language shall be some json file that is the input for other tools. 
Some internal technical stuff needs to be configured (not compile-time-fixed but not part of the dsl). 
What is the IntelliJ way to do such stuff? Yes... Preferences. MPS has support for it.
Inside the JSON-textGen I need access to these Project-Plugin-Preferences. And the only MPS-way is via the Project.

As the only input to the textGen is a node, we have to guess.
In TexGen none of the documented options seems to work (all returned null). 

What we had done: A reverse scan across all opens (not disposed) projects. 
Then we walked across all modules and root-nodes. At least the nodes seem the have to same ids as in the source-model. Models and Modules seems to be temporary created (I guess this is how MPS works...).

```
public static Project getProject(SNode snode) {
   ... tries all official ways to get it (fails from textGen)... 
  //  find some info what is the replacement for this deprecated getInstance...
  ProjectManager pm = ProjectManager.getInstance(); 
  for (Project p : pm.getOpenedProjects()) { 
    if (!p.isDisposed()) { 
      for (SModule md : p.getProjectModules()) { 
        //  Seems the only reliable way...  scan for the node-id,   
        //  all other stuff around (models, modules) seem to be temporary copies.  
        for (SModel rm : md.getModels()) { 
          for (SNode nd : rm.getRootNodes()) { 
            if (nd.getNodeId().equals(snode.getNodeId())) { 
              return p; 
   } } } } } }
   return null;
}
```


MPS shares modules across all open projects... But the node should only reached this by the containing project. 
As we plan to use this code from a CI-pipeline, a scan across "Open Windows" will not work (mps REALLY works headless, we tested it).

Any critic or enlightenment is welcome...

**Reply by alexanderpann** [reply]
I have experimented with accessing the project from TextGen and came up with the following code that works at least locally:
```
MPSModuleRepository repository = MPSCoreComponents.getInstance().getPlatform().findComponent(MPSModuleRepository.class); 
MPSNodeVirtualFile file = NodeVirtualFileSystem.getInstance().getFileFor(repository, rootNode); 
Project project = ProjectLocator.getInstance().guessProjectForFile(file);
```

Does this solve your issue?

**Reply by BWeng20** [reply]
Yes, I can try.  Looks good.

I am not sure if we will have a "file". I saw during textGen, that the nodes were from an in-memory-copy of the model (I have no clue why, I guess that's how MPS works). I will try it, monday at work :)

**Reply by BWeng20** [reply]
Yes, it works, 
Thank you very much!

Best Regards,
Bernd
